manager/allocator/network/ipam/ipam.go

```diff
@@ -0,0 +1,729 @@
+
+	"github.com/docker/swarmkit/api"
+)
+
+// DrvRegistry is an interface defining
+type DrvRegistry interface {
+	IPAM(name string) (ipamapi.Ipam, *ipamapi.Capability)
+}
+
+// network is a private type used to hold the internal state of a network in
+// the IP Allocator
+type network struct {
+	// nw is a local cache of the network object found in the store
+	nw *api.Network
+
+	// pools is used to save the internal poolIDs needed when releasing a pool.
+	// It maps an ip address to a pool ID.
+	pools map[string]string
+
+	// endpoints is a map of the endpoint IP to the poolID from which it was
+	// allocated.
+	endpoints map[string]string
+}
+
+// Allocator is an interface that represents the IP address allocator. It
+// exists mainly for testing purposes, so that the allocator can be more easily
+// mocked out.
+type Allocator interface {
+	Restore([]*api.Network, []*api.Endpoint, []*api.NetworkAttachment) error
+	AllocateNetwork(*api.Network) error
+	DeallocateNetwork(*api.Network)
+	AllocateVIPs(*api.Endpoint, []string) error
+	DeallocateVIPs(*api.Endpoint)
+	AllocateAttachments([]*api.NetworkAttachmentConfig) ([]*api.NetworkAttachment, error)
+	DeallocateAttachments([]*api.NetworkAttachment)
+}
+
+// allocator is an allocator for IP addresses and IPAM pools. It handles all
+// allocation and deallocation of IP addresses, both for VIPs and endpoints, in
+// addition to handling IPAM pools for networks.
+type allocator struct {
+	// networks maps network IDs to the locally stored network object
+	networks map[string]*network
+
+	// drvRegistry is the driver registry, from which we can get active network
+	// drivers
+	drvRegistry DrvRegistry
+}
+
+// NewAllocator takes a drvRegistry and creates a new IP allocator
+func NewAllocator(reg DrvRegistry) Allocator {
+	return &allocator{
+		networks:    make(map[string]*network),
+		drvRegistry: reg,
+	}
+}
+
+// Restore restores the state of the provided networks to the Allocator. It can
+// return errors if the initialization of its dependencies fails.
+func (a *allocator) Restore(networks []*api.Network, endpoints []*api.Endpoint, attachments []*api.NetworkAttachment) error {
+	// Initialize the state of the networks. this gets the initial IPAM pools.
+	for _, nw := range networks {
+		// if the network has no IPAM field, it has no state, and there is
+		// nothing to do
+		if nw.IPAM == nil ||
+			// if the network has no IPAM driver, it has no IPAM state, and there
+			// is nothing to do.
+			nw.IPAM.Driver == nil ||
+			// if the network has no ipam configs, then it has no state, and there
+			// is nothing to do
+			len(nw.IPAM.Configs) == 0 {
+			continue
+		}
+		local := &network{
+			nw:        nw,
+			pools:     make(map[string]string),
+			endpoints: make(map[string]string),
+		}
+		ipamName := nw.IPAM.Driver.Name
+		ipamOpts := nw.IPAM.Driver.Options
+		if ipamOpts == nil {
+			ipamOpts = map[string]string{}
+		}
+		// if we have an ipam driver name and we have IPAM configs, but for
```

Comment:
Non-blocking - would it make sense to just assign this in the block that specifies what to do if `ipamOpts` are `nil`?  I don't think it makes a difference to reassign back if it's not `nil`, then you may not need a comment explaining it.

Reply:
i honestly don't know why i did it like this 🤔 🤔 🤔